README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->
LexisPlotR
==========

[![CRAN](http://www.r-pkg.org/badges/version/LexisPlotR)](http://cran.rstudio.com/package=LexisPlotR) [![Downloads](http://cranlogs.r-pkg.org/badges/LexisPlotR?color=brightgreen)](http://www.r-pkg.org/pkg/LexisPlotR) [![Travis build status](https://travis-ci.org/ottlngr/LexisPlotR.svg?branch=master)](https://travis-ci.org/ottlngr/LexisPlotR) [![AppVeyor build status](https://ci.appveyor.com/api/projects/status/github/ottlngr/LexisPlotR?branch=master&svg=true)](https://ci.appveyor.com/project/ottlngr/LexisPlotR)

`LexisPlotR` is a tool to easily plot Lexis Diagrams within R. It is based on [`ggplot2`](https://github.com/hadley/ggplot2) and wraps necessary steps like drawing the grid, highlighting certain areas or inserting lifelines in a couple of easy to use functions.

### What is a Lexis Diagram?

> In demography a Lexis diagram (named after economist and social scientist Wilhelm Lexis) is a two dimensional diagram that is used to represent events (such as births or deaths) that occur to individuals belonging to different cohorts. Calendar time is usually represented on the horizontal axis, while age is represented on the vertical axis. (<https://en.wikipedia.org/wiki/Lexis_diagram>)

### Major changes from v0.3.\* to v0.4

With v0.4 a complete rewrite of `LexisPlotR` started. To prevent existing code from failing, v0.4 still keeps functions from earlier versions, but only introduces new functionality in functions named using 'snake\_case'. So if you are new to `LexisPlotR` do not use functions named using dot notation (e.g. `lexis.grid()`) but functions using snake\_case (e.g. `lexis_grid()`). If you have existing code using functions from earlier versions of `LexisPlotR` please adapt new function names and syntax where necessary.

### Installation

You can install the latest version of **`LexisPlotR`** by using `install_github()` from the `devtools`-package:

    devtools::install_github("ottlngr/LexisPlotR")
    library(LexisPlotR)

`LexisPlotR` (v0.4) is available on CRAN. Install v0.4 from CRAN:

    install.packages("LexisPlotR")

### Using LexisPlotR

`LexisPlotR` provides a couple of functions to draw Lexis Diagrams in R. Besides the ability to draw empty Lexis grids, `LexisPlotR` also offers some functionality to highlight certain areas of the grid or to add actual data to the Lexis Diagram.

##### Plot an empty Lexis grid

A Lexis Diagram is basically determined by two measures: A range of years presented on the horizontal axis and a range of ages shown on the vertical axis. To plot an empty Lexis grid, use `lexis_grid()` which takes these measures as numeric inputs:

``` r
library(LexisPlotR)
```

``` r
# Plot a Lexis grid from year 1900 to year 1905, representing the ages from 0 to 5
lexis_grid(year_start = 1900, year_end = 1905, age_start = 0, age_end = 5)
```

![](man/figures/unnamed-chunk-3-1.png)

The aspect ratio of the axes is fixed to ensure right-angled triangles. So even non-square Lexis grids show right-angled triangles:

``` r
lexis_grid(year_start = 1900, year_end = 1905, age_start = 0, age_end = 8)
```

![](man/figures/unnamed-chunk-4-1.png)

When using wide age and/or date ranges, it may be helpful to define a `delta` specifying the number of years to group in the grid:

``` r
lexis_grid(year_start = 1900, year_end = 1950, age_start = 0, age_end = 50, delta = 5)
```

![](man/figures/unnamed-chunk-5-1.png)

##### Highlight certain areas of the Lexis grid

Sometimes it is useful to highlight certain areas in the Lexis Diagram, like a certain age, year or cohort.

Highlighting a certain age in your grid is supported by `lexis_age` which will draw a coloured rectangle inside your grid marking all points in the grid belonging to a certain age group.

``` r
lexis <- lexis_grid(year_start = 1900, year_end = 1905, age_start = 0, age_end = 5)
lexis_age(lg = lexis, age = 2)
```

![](man/figures/unnamed-chunk-6-1.png)

Next to `lexis_age()` there are also `lexis_year()` and `lexis_cohort()` which highlight a certain year or cohort, respectively:

``` r
lexis <- lexis_grid(year_start = 1900, year_end = 1905, age_start = 0, age_end = 5)
lexis_year(lg = lexis, year = 1903)
```

![](man/figures/unnamed-chunk-7-1.png)

``` r
lexis <- lexis_grid(year_start = 1900, year_end = 1905, age_start = 0, age_end = 5)
lexis_cohort(lg = lexis, cohort = 1898)
```

![](man/figures/unnamed-chunk-8-1.png)

The similar `lexis_polygon()` can be used to highlight arbitrary regions of the Lexis grid:

``` r
lexis <- lexis_grid(year_start = 1900, year_end = 1905, age_start = 0, age_end = 5)

polygons <- data.frame(group = c(1, 1, 1, 2, 2, 2),
                       x = c("1901-01-01", "1902-01-01", "1902-01-01", "1903-01-01", "1904-01-01", "1904-01-01"),
                       y = c(1, 1, 2, 1, 1, 2))

lexis_polygon(lg = lexis, x = polygons$x, y = polygons$y, group = polygons$group)
```

![](man/figures/unnamed-chunk-9-1.png)

##### Add life lines to the Lexis Diagram

A life line is a simple tool to represent an individual's life in a Lexis Diagram.

The behaviour of `lexis_lifeline()` can be controlled by supplying an entry and/or exit date along with the individuals birthday:

``` r
lg <- lexis_grid(year_start = 1900, year_end = 1905, age_start = 0, age_end = 5)
lexis_lifeline(lg = lg, birth = "1901-09-23", lwd = 1.5)
```

![](man/figures/unnamed-chunk-10-1.png)

``` r
lexis_lifeline(lg = lg, birth = "1901-09-23", entry = "1902-04-01", lwd = 1.5)
```

![](man/figures/unnamed-chunk-10-2.png)

``` r
lexis_lifeline(lg = lg, birth = "1901-09-23", entry = "1902-04-01", exit = "1904-10-31", lwd = 1.5)
```

![](man/figures/unnamed-chunk-10-3.png)

You can also use entry and death dates from a `data.frame` which is useful when plotting life lines of several individuals or hole populations. **`LexisPlotR`** comes with a random dataset of entry and exit dates for 300 Individuals from 1895 to 1905. Some of the deaths (or exits) are not observed or unknown. Take a look at the `lifelines_sample` dataset:

``` r
data("lifelines_sample")
lifelines_sample <- lifelines_sample[1:30,]
head(lifelines_sample, 10)
```

    ##         entry       exit
    ## 1  1898-04-25 1898-07-30
    ## 2  1899-12-28       <NA>
    ## 3  1903-01-15       <NA>
    ## 4  1901-04-13       <NA>
    ## 5  1895-05-30 1900-03-29
    ## 6  1897-09-22       <NA>
    ## 7  1896-02-16 1896-04-24
    ## 8  1896-11-13 1902-10-30
    ## 9  1904-10-31       <NA>
    ## 10 1899-04-02 1902-04-11

To add all this data to your Lexis Diagram, use `lexis.lifeline()` and provide the respective columns of `lifelines_sample` as arguments:

``` r
lg <- lexis_grid(year_start = 1900, year_end = 1905, age_start = 0, age_end = 5)
lexis_lifeline(lg = lg, birth = lifelines_sample$entry, exit = lifelines_sample$exit, lwd = 1.5, lineends = TRUE)
```

![](man/figures/unnamed-chunk-12-1.png)

#### Chaining `LexisPlotR` functions

`LexisPlotR` functions can be chained using the pipe operator `%>%` which can be imported from `magrittr`, `dplyr` and other packages.

``` r
library(magrittr)

lexis_grid(year_start = 1990, year_end = 2000, age_start = 0, age_end = 10) %>%
  lexis_age(age = 3) %>%
  lexis_cohort(cohort = 1988) %>%
  lexis_year(year = 1998) %>%
  lexis_lifeline(birth = "1991-09-23", lwd = 1.5)
```

![](man/figures/unnamed-chunk-13-1.png)

#### `LexisPlotR` is build on top of `ggplot2`

`LexisPlotR` is actually just a collection of wrapper functions encapsulating `ggplot2` logic. Therefore, each plot returned by `LexisPlotR` is an object of class `ggplot` and can be altered using well known `ggplot2` functionalities.

``` r
library(magrittr)
library(ggplot2)

p <- lexis_grid(year_start = 1990, year_end = 2000, age_start = 0, age_end = 10) %>%
  lexis_age(age = 3) %>%
  lexis_cohort(cohort = 1988) %>%
  lexis_year(year = 1998) %>%
  lexis_lifeline(birth = "1991-09-23", lwd = 1.5)

p <- p + labs(x = "Age", y = "Year", "title" = "LexisPlotR")
p
```

![](man/figures/unnamed-chunk-14-1.png)

#### Use public available data within LexisPlotR

Projects like the [Human Mortality Database (HMD)](https://mortality.org) provide demographic data for many countries in the world.

##### HMD

HMD's Lexis Triangle data can be visualized using `LexisPlotR`'s `lexis_polygon()`. To fetch data from HMD, the [`HMDHFDplus`](https://github.com/timriffe/TR1/tree/master/TR1/HMDHFDplus) package can be used:

``` r
library(LexisPlotR)
library(ggplot2)
library(HMDHFDplus)

triangles <- HMDHFDplus::readHMDweb(CNTRY = "RUS", "Deaths_lexis", Sys.getenv("HMD_USER"), Sys.getenv("HMD_PASSWORD"))

tidy_triangles <- tidy_triangle_data(triangles)

lg <- lexis_grid(1960, 2000, 30, 80, delta = 5)

lg + geom_polygon(data = tidy_triangles, aes(x = x, y = y, group = id, fill = Total))
```

##### Other sources for demographic data

Besides HMD, there are other similar or related projects that provide demographic data that could be used within LexisPlotR. An incomplete list:

-   [Japanese Mortality Database](http://www.ipss.go.jp/p-toukei/JMD/index-en.asp)
-   [Canadian Human Mortality Database](http://www.bdlc.umontreal.ca/CHMD/)
-   [Human Fertility Database](https://www.humanfertility.org/cgi-bin/main.php)
-   ...

Until now, only HMD Lexis Triangles are supported natively by LexisPlotR.


NEWS.md
# LexisPlotR 0.4

- Many minor improvements.
- Unified functions and arguments, changed function naming scheme from `lexis.grid()` to `lexis_grid()`.
- Set legacy functions to `.Deprecated`.
- New feature: `lexis_lifeline()` now lets start and end lifelines at arbitrary points in time.
- New feature: `lexis_polygon()` can draw arbitrary polygons inside a Lexis grid.
- Improved support for HMD Lexis Triangle data.

R/lexisplotr.R
#' Plot a Lexis Diagram.
#' 
#' lexisplotr is an easy to use function to plot Lexis Diagrams.
#' 
#' 
#' @param from_year integer, set the year the Lexis Diagram starts with.
#' @param to_year integer, set the year the Lexis Diagram ends with.
#' @param from_age integer, set the age the Lexis Diagram starts with.
#' @param to_age integer, set the age the Lexis Diagram ends with.
#' @param cohort integer, set a cohort to emphasize.
#' @param year integer, set a year to emphasize.
#' @param age integer, set an age to emphasize.
#' @param lifelines list, set lifelines to be drawn. The list contains each
#' lifeline as a vector specifying the date of the beginning of the observation 
#' and the date of the end of the observation. Dates in format \code{"dd.mm.yyyy"}.
#' @param triangle list, set triangles to be drawn. The list contains each lifeline
#' as a vector of a year-age-combination and a keyword to determine whether the upper \code{"u"}
#' or lower \code{"l"} triangle shall be drawn.
#' @param colors vector, set the colors used for \code{cohort, year, age, lifelines, triangle}.
#' Default: \code{colors=c("yellow", "green", "blue", "red", "purple")}
#' @param title character, title of the plot. Default: "Lexis Diagram".
#' @param envir, Default: \code{environment()}. Should not be changed. Needed as a workarround.
#' @details Required date format: \code{"dd.mm.yyyy"}.
#' 
#' The function determines the aspect ratio of the x- and y-axis to enforce
#' isosceles triangles. The aspect ratio will not be effected by defining
#' \code{width} and \code{height} in \code{pdf()} or other graphic devices.
#' 
#' Because the returned object is a ggplot2 graph, the overall appearence of
#' the graph can be edited by adding e.g. \code{themes()} to the plot.
#' @return The functions returns a ggplot2-plot.
#' @note To save the plot, using \code{pdf()} is recommended since it will
#' achieve good results and avoid overlapping in most cases. Unix users may use
#' \code{x11()} to start a new graphical device to preview the plot.
#' @author Philipp Ottolinger
#' @seealso \code{\link{ggplot}}
#' @keywords hplot
#' @export lexisplotr
#' @import ggplot2
#' @examples
#' library(LexisPlotR)
#' 
#' ## Plot a Lexis Diagram grid
#' lexisplotr(from_year=1900, to_year=1905, from_age=0, to_age=5)
#' 
#' ## Emphasize a certain cohort, year and age
#' lexisplotr(from_year=1900, to_year=1905, from_age=0, to_age=5,
#'            cohort=1901, year=1903, age=2)
#'            
#' ## Plot lifelines
#' lexisplotr(from_year=1900, to_year=1905, from_age=0, to_age=5,
#'            lifelines=list(c("30.06.1900", "23.09.1903"),
#'                           c("01.01.1901", "31.12.1904")))
#'                           
#' ## Emphasize a certain triangles
#' lexisplotr(from_year=1900, to_year=1905, from_age=0, to_age=5,
#'            triangle=list(c(1901,2,"u"),c(1902,4,"l")))
#'           
#' ## Change the size of the axis text
#' lexis <- lexisplotr(from_year=1900, to_year=1905, from_age=0, to_age=5)
#' lexis <- lexis + theme(axis.text = element_text(size=10))
#' lexis
lexisplotr <- function(from_year, 
                       to_year, 
                       from_age, 
                       to_age, 
                       age, 
                       cohort, 
                       year, 
                       triangle, 
                       lifelines,
                       colors=c("yellow", "green", "blue", "red", "purple"),
                       title="Lexis Diagram",
                       envir=environment()){
  
  #### Create data ####
  from_year.d <- as.Date(paste("01.01.", from_year, sep=""), "%d.%m.%Y")
  to_year.d <- as.Date(paste("01.01.", to_year, sep=""),"%d.%m.%Y")
  years <- from_year:to_year
  years.d <- seq(from_year.d, to_year.d, "year")
  ages <- from_age:to_age
  
  
  #### Plot grid ####
  lex <- ggplot() +
    geom_abline(
      aes(
        intercept=-as.numeric(from_year.d)/364.25+seq(-to_age+1,to_age-1,1), 
        slope=1/364.25),
      lwd=0.4) + 
    geom_hline(
      yintercept=seq(from_age,to_age,1),
      lwd=0.4) +
    geom_vline(
      xintercept = as.numeric(years.d),
      lwd=0.4) +
    scale_x_date(
      limits=c(from_year.d, to_year.d), 
      date_breaks="1 year", 
      expand=c(0,0),
      date_labels="%Y",
      name="Year") +
    scale_y_continuous(
      limits=c(from_age, to_age), 
      expand=c(0,0),
      breaks=ages,
      name="Age")
  
  #### Appearance ####
  lex <- lex +
    theme_bw() +
    theme(
      aspect.ratio=(to_age-from_age)/(as.numeric(to_year.d)/364.25-as.numeric(from_year.d)/364.25),
      axis.line = element_line(colour="black", size=0.3),
      axis.ticks = element_line(colour="black"),
      axis.text.x = element_text(size=14, colour="black"),
      axis.text.y = element_text(size=14, colour="black"),
      axis.title = element_text(size=18, face="bold"),
      plot.title = element_text(size=25, face="bold", vjust=1.5),
      panel.grid.major.y = element_blank(),
      panel.background = element_rect(fill='white', colour='black'),
      panel.grid.major.x = element_blank()) +
    ggtitle("Lexis Diagram")
  
  #### Plot age ####
  if (!missing(age)) {
    lex <- lex + 
      geom_polygon(
        aes(x=c(from_year.d, to_year.d, to_year.d, from_year.d), 
            y=c(age,age, age+1, age+1)), 
        fill=colors[1], 
        col=colors[1], 
        alpha=0.5)
  }
  #### plot cohort ####
  if (!missing(cohort)) {
    lex <- lex + 
      geom_polygon(
        aes(x=c(as.Date(paste("01.01.", cohort, sep=""), "%d.%m.%Y"),
                as.Date(paste("01.01.", cohort+1, sep=""), "%d.%m.%Y"),
                to_year.d, 
                to_year.d),
            y=c(from_age, 
                from_age, 
                to_year - cohort-1, 
                to_year - cohort)),
        col=colors[2], 
        fill=colors[2], 
        alpha=0.5)
  }
  #### plot year ####
  if (!missing(year)) {
    lex <- lex + 
      geom_polygon(
        aes(
          x=c(as.Date(paste("01.01.", year, sep=""), "%d.%m.%Y"),
              as.Date(paste("01.01.", year+1, sep=""), "%d.%m.%Y"),
              as.Date(paste("01.01.", year+1, sep=""), "%d.%m.%Y"),
              as.Date(paste("01.01.", year, sep=""), "%d.%m.%Y")),
          y=c(from_age, from_age, to_age, to_age)),
        col=colors[3], 
        fill=colors[3], 
        alpha=0.5)
  }
  #### plot triangles ####
  #triangle <- list(c(1988,2,"u"),c(1988,3,"u"))
  if (!missing(triangle)) {
    for (i in 1:length(triangle)) {
      if (triangle[[i]][3] == "u") {
        lex <- lex + 
          geom_polygon(
            aes_string(
              x=c(as.Date(paste("01.01.", as.numeric(triangle[[i]][1]), sep=""), "%d.%m.%Y"),
                  as.Date(paste("01.01.", as.numeric(triangle[[i]][1])+1, sep=""), "%d.%m.%Y"),
                  as.Date(paste("01.01.", as.numeric(triangle[[i]][1]), sep=""), "%d.%m.%Y")),
              y=c(as.numeric(triangle[[i]][2]),
                  as.numeric(triangle[[i]][2])+1,
                  as.numeric(triangle[[i]][2])+1)),
            col=colors[4],
            fill=colors[4],
            alpha=0.5
          )
      }
      if (triangle[[i]][3] == "l") {
        lex <- lex + 
          geom_polygon(
            aes_string(
              x=c(as.Date(paste("01.01.", as.numeric(triangle[[i]][1]), sep=""), "%d.%m.%Y"),
                  as.Date(paste("01.01.", as.numeric(triangle[[i]][1])+1, sep=""), "%d.%m.%Y"),
                  as.Date(paste("01.01.", as.numeric(triangle[[i]][1])+1, sep=""), "%d.%m.%Y")),
              y=c(as.numeric(triangle[[i]][2]),
                  as.numeric(triangle[[i]][2]),
                  as.numeric(triangle[[i]][2])+1)), 
            col=colors[4],
            fill=colors[4],
            alpha=0.5
          )
      }
    }
  }
  #### Plot lifelines ####
  # lifelines <- list(c("01.06.1987","01.06.1993"))
  if (!missing(lifelines)) {
    for (i in 1:length(lifelines)) {
      ll <- lifelines[[i]]
      lex <- lex + 
        geom_segment(
          aes_string(
            x=as.Date(ll[1], "%d.%m.%Y"),
            xend=as.Date(ll[2], "%d.%m.%Y"),
            y=0,
            yend=as.numeric(as.Date(ll[2], "%d.%m.%Y")-as.Date(ll[1], "%d.%m.%Y"))/364.25),
          col=colors[5],
          lwd=1
        )
    }
  }
  #### return ####
  return(lex)
}

#### Test call ####
# lexisplotr(
#   from_year=1987,
#   to_year=1994,
#   from_age=0,
#   to_age=7, 
#   age=4, 
#   cohort=1988, 
#   year=1991, 
#   triangle=list(c(1987,2,"u"),c(1992,1,"l"), c(1989,5,"u")), 
#   lifelines = list(c("01.06.1987","01.06.1993"), c("01.02.1991","30.11.1993"))
# )